x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java

```diff
@@ -344,18 +350,100 @@ private void createNewProfile(Subject subject, ActionListener<Profile> listener)
                 bulkRequest,
                 TransportSingleItemBulkWriteAction.<IndexResponse>wrapBulkResponse(ActionListener.wrap(indexResponse -> {
                     assert docId.equals(indexResponse.getId());
-                    // TODO: replace with actual domain information
                     final VersionedDocument versionedDocument = new VersionedDocument(
                         profileDocument,
                         indexResponse.getPrimaryTerm(),
                         indexResponse.getSeqNo()
                     );
                     listener.onResponse(versionedDocument.toProfile(Set.of()));
-                }, listener::onFailure))
+                }, e -> {
+                    if (e instanceof VersionConflictEngineException) {
+                        // Document already exists with the specified ID, get the document with the ID
+                        // and check whether it is the right profile
+                        getVersionedDocument(profileDocument.uid(), ActionListener.wrap(versionedDocument -> {
```

Comment:
Instead of the get doc here, it _might_ be simpler to reason about the whole logic if instead we redo the search.
The function that does the search (some internal function of `getVersionedDocument`), can take a parameter for the document id that it tries to use to create docs if the search returns empty. If the create doc fails, the method calls itself recursively with the incremented doc id.

Reply:
We do need improve the robustness for retrieving relevant documents after seeing a VersionConflict. Neither Search or Get can guarantee to return the necessary document in all cases (Get has a better chance because it is realtime but still not perfect). It's likely we need some backoff and retry strategy here. I will work on it with your suggestion in mind.